server/authentication.js

```diff
@@ -8,42 +13,220 @@ function enforceHTTPS(request, response, next) {
   return next();
 }
 
-// Only allow folks who are authorized.  Call `next` if they are authorized,
-// and return a 404 with no body if not.
-function onlyAllowResearchers(request, response, next) {
-  // TODO
-  // 1. Read the token from the request header
-  // 2. Check to see if the token header is in `tokens` database table and it's less than 24 hours old
+function getDomain(request) {
+  return (process.env.NODE_ENV === 'development')
+    ? 'http://localhost:3000'
+    : `https://${request.headers.host}`;
+}
+
+// Middleman function to confirm authorization token is valid
+// Reads token from request header and checks against tokens in db
+function onlyAllowResearchers(pool, request, response, next) {
+  //TODO: Is this really true?
   // 3. This code should run in both development and production.
-  if (process.env.NODE_ENV === 'development') return next();
-  response.status(404);
-  response.end();
+
+  const token = request.headers['token'];
+  const email = request.headers['email'];
+
+  checkToken(pool, email, token)
+    .then(tokenAuthorized => {
+      console.log('tokenAuthorized:',tokenAuthorized);
+      if (tokenAuthorized) {
+        return next();
+      }
+      else if (tokenAuthorized===false){
+        console.log('token is incorrect');
+        return response.status(405).end();
+      }
+      else {
+        console.log('something went wrong');
+        return response.status(500).end();
+      }
+    })
+    .catch(err => {
+      console.log({ error: err });
+      return response.status(500).end();
+    });
 }
 
+function checkToken(pool, email, token) {
+  const now = new Date();
 
-function loginEndpoint(req, res){
-  // TODO:
-  // 1. Check that the email is in the whitelist in the `researchers` database table
-  // 2. Insert a new record to the `links` database table
-  // 3. Send an email to the researcher with that link in it (generate email with Mustache template, send it with Mailgun)
-  // 4. Return a 200 response
-  res.status(405);
-  res.end();
+  const sql = `
+    SELECT * 
+    FROM tokens 
+    WHERE token=$1 
+      AND email=$2
+      AND $3 > timestampz
+      And $3 < (timestampz + INTERVAL '24 hours')
+    ORDER BY id ASC LIMIT 1`;
+  const values = [token, email, now];
+  return pool.query(sql, values)
+    .then(results => Promise.resolve(results.rowCount===1))
+    .catch(err => {
+      console.log('query returned err: ', err);
```

Comment:
It'd be good to make this error message make sense on its own (eg, "checkToken query returned err trying to look up a login token") or something that so when you read it as a log message, you know where it's coming from and the first step of what it might mean.  Right now you might be like "what query?" and then just have to search the code for it.

Reply:
Makes sense. I'll update all of the catch error messages.